napari/utils/translations.py

```diff
@@ -166,11 +167,99 @@ def get_language_packs(display_locale: str = _DEFAULT_LOCALE) -> dict:
 
 # --- Translators
 # ----------------------------------------------------------------------------
+class TranslationString:
+    """
+    TODO:
+    """
+
+    def __init__(
+        self,
+        domain: str = None,
+        msgctxt: str = None,
+        msgid: str = None,
+        msgid_plural: str = None,
+        n: str = None,
+        deferred: bool = False,
+        **kwargs,
+    ):
+        self._domain = domain
+        self._msgctxt = msgctxt
+        self._msgid = msgid
+        self._msgid_plural = msgid_plural
+        self._n = n
+        self._deferred = deferred
+        self._kwargs = kwargs
+
+        # Add `n` to `kwargs` to use with `format`
+        self._kwargs['n'] = n
+
+    def __repr__(self):
+        """"""
+        return repr(self.__str__())
+
+    def __str__(self):
+        """"""
+        return self.value() if self._deferred else self.translation()
+
+    def value(self):
+        """
+        Return the original string with interpolated kwargs, if provided.
+        """
+        if self._n is None or self._n == 1:
+            string = self._msgid
+        elif self._n != 1:
+            string = self._msgid_plural
+
+        return string.format(**self._kwargs)
+
+    def translation(self):
+        """
+        Return the translated string with interpolated kwargs, if provided.
+        """
+        # Python 3.7 or lower does not offer translations based on context.
+        # On these versions `gettext.npgettext` falls back to `gettext.ngettext`
+        if PY37_OR_LOWER:
+            if self._n is None:
+                translation = gettext.dgettext(self._domain, self._msgid)
+            else:
+                translation = gettext.dngettext(
+                    self._domain, self._msgid, self._msgid_plural, self._n
+                )
+        else:
+            if self._n is None and self._msgctxt is None:
```

Comment:
I have little worries about the number of ifs in these functions.  

These two lines could be merged in one `elif` but it may break semantic code split but I would like to see something simpler here.

Reply:
Yes, probably, but keeping it separate makes it clear to understand what is going on.

Reply:
i'm ok leaving this as its a little more readable, but thanks for flagging for extra discussion

Reply:
maybe split this function and simplify all code using fact that there is no need for `else` after the `return` statement.

Reply:
I personally do not like having multiple return statements in a single function, specially if it is a longer one like this one.

```python

def something():
    if some_condition:
        return 

    if some_condition:
        return 

    if some_condition:
        return 

    if some_condition:
        return 

    if some_condition:
        return 

    if some_condition:
        return 

    if some_condition:
        return 

    if some_condition:
        return

Reply:
I personally prefer multiple `return` statements. Reading function I see that this is the end of branch and I do not need this part of code to understand the next part.

Reply:
This feels like a bit of a bikeshed discussion.  I find both reasonable in various cases, and lacking a general style guide for napari, I think it’s fine to leave it up to the contributor of the PR

Reply:
Agreed :)